### ISSUE TITLE
Scorpio classic crashes when a large user buffer is used to get a variable

### ISSUE THREAD

**Issue by dqwu** [E3SM-Project/scorpio#478]
This issue can be reproduced with a simple E3SM ne4 F case using scorpio classic (not reproducible with scorpio):
`pio_support::pio_die:: myrank=          -1 : ERROR: pionfget_mod.F90:         432 : Input/Output data amount mismatch`

In one of the intermediate changes made in E3SM-Project/E3SM#4379 , the user provided a large buffer (than the size of the variable) when getting a variable from the input file.

For pnetcdf type, the following API call would fail when isize (calculated from user buffer ival, e.g. 51) is larger than the actual variable size (e.g. 25):
`ierr = nfmpi_get_var_all(File%fh, varid, ival, isize, {MPITYPE})`

The returned error code is:
`NF_EIOMISMATCH = -209   ! Input/Output data amount mismatch`
To fix this error, the argument isize needs to be replaced with the actual variable size.

For netcdf/netcdf4c/netcdf4p type, the following API call would fail when the implicit start/count (calculated from user buffer ival by NetCDF lib) exceeds dimension bound of the variable:
`ierr = nf90_get_var(File%fh, varid, ival)`

The returned error code is:
`NF_EEDGE = -57   ! Start+count exceeds dimension bound`
To fix this error, explicit start/count needs to be provided to get all data of the variable.

A related fix for scorpio is PR #417

**Comment by dqwu** [comment]
@jayeshkrishna FYI, a quick and dirty fix will be something like:
```
--- a/pio/pionfget_mod.F90.in
+++ b/pio/pionfget_mod.F90.in
@@ -358,6 +358,7 @@ CONTAINS
 !<
   integer function get_var_{DIMS}d_{TYPE} (File,varid, ival) result(ierr)
     use pio_msg_mod, only : pio_msg_getvar_{DIMS}d
+    use nf_mod, only : pio_inq_varndims, pio_inq_vardimid, pio_inq_dimlen
     type (File_desc_t), intent(in) :: File
     integer, intent(in) :: varid
     {VTYPE}, intent(out) :: ival{DIMSTR}
@@ -369,6 +370,13 @@ CONTAINS
     integer :: i
 #endif
     integer(kind=PIO_OFFSET) :: isize
+    integer(kind=PIO_OFFSET) :: var_sz
+    integer :: ndims = 0
+    integer :: j
+    integer, allocatable :: dimids(:)
+    integer, allocatable :: dim_sz(:)
+    integer, allocatable :: my_start(:)
+    integer, allocatable :: my_count(:)
 
 #ifdef TIMING
     call t_startf("PIO:pio_get_var_{DIMS}d_{TYPE}")
@@ -405,21 +413,74 @@ CONTAINS
 
     endif
 
+    ! Get the number of dimensions in variable
+    ierr = pio_inq_varndims(File, varid, ndims)
+    if(ierr /= PIO_NOERR) then
+       call piodie(__PIO_FILE__, __LINE__, "Inquiring number of dims of variable failed")
+    end if
+
+    ! Get the dimension ids for the variable
+    if(ndims == 0) then
+       ! A scalar variable
+       var_sz = 1
+
+       allocate(my_start(1))
+       allocate(my_count(1))
+       my_start(1) = 1
+       my_count(1) = 1
+    else
+       allocate(dimids(ndims))
+       ierr = pio_inq_vardimid(File, varid, dimids)
+       if(ierr /= PIO_NOERR) then
+          call piodie(__PIO_FILE__, __LINE__, "Inquiring the variable dimension ids for the variable failed")
+       end if
+
+       allocate(dim_sz(ndims))
+       allocate(my_start(ndims))
+       allocate(my_count(ndims))
+
+       var_sz = 1
+       do j=1,ndims
+          ierr = pio_inq_dimlen(File, dimids(j), dim_sz(j))
+          if(ierr /= PIO_NOERR) then
+             call piodie(__PIO_FILE__, __LINE__, "Inquiring the lengths of dimension for the variable failed")
+          end if
+          var_sz = var_sz * INT(dim_sz(j), PIO_OFFSET_KIND)
+          my_start(j) = 1
+          my_count(j) = dim_sz(j)
+       end do
+
+       deallocate(dim_sz)
+       deallocate(dimids)
+    end if
 
+    if(isize < var_sz) then
+       print *, "ERROR: The size of the user buffer (", isize, " elements) ",&
+                 "is smaller than the size of the variable (", var_sz, " elements)"
+       call piodie(__PIO_FILE__, __LINE__, "Insufficient user buffer size")
+    end if
 
     if(File%iosystem%IOProc) then
        select case (iotype) 
 #ifdef _PNETCDF
        case(pio_iotype_pnetcdf)
-          ierr = nfmpi_get_var_all(File%fh, varid, ival, isize, {MPITYPE})
+          ierr = nfmpi_get_var_all(File%fh, varid, ival, var_sz, {MPITYPE})
 #endif
 #ifdef  _NETCDF
        case(pio_iotype_netcdf4p)
+#if ({DIMS} > 0)
+             ierr = nf90_get_var(File%fh, varid, ival, my_start, my_count)
+#else
              ierr = nf90_get_var(File%fh, varid, ival)
+#endif
        case(pio_iotype_netcdf, pio_iotype_netcdf4c)
           ! Only io proc 0 will do reading
           if (File%iosystem%io_rank == 0) then
+#if ({DIMS} > 0)
+             ierr = nf90_get_var(File%fh, varid, ival, my_start, my_count)
+#else
              ierr = nf90_get_var(File%fh, varid, ival)
+#endif
           end if
           if(.not. ios%async_interface .and. ios%num_tasks==ios%num_iotasks) then
              call MPI_BCAST(ival,int(isize), {MPITYPE} ,0,ios%IO_comm, mpierr)
@@ -429,6 +490,8 @@ CONTAINS
 #endif
        end select
     end if
+    deallocate(my_start)
+    deallocate(my_count)
     call check_netcdf(File,ierr,__PIO_FILE__,__LINE__)
     if(ios%async_interface .or. ios%num_tasks>ios%num_iotasks) then
        call MPI_Bcast(ival,int(isize), {MPITYPE} , ios%IOMaster, ios%My_comm, mpierr)
```

**Comment by wkliao** [comment]
Actually, you can simply set `isize` to -1, i.e.
```
ierr = nfmpi_get_var_all(File%fh, varid, ival, -1, {MPITYPE})
```
Using -1 means the number of get elements matches the argument `count`.
In this `var` API case, it is the whole variable.

The requirement is that `MPITYPE` must be the I/O buffer's data type,
an MPI predefined datatype, I assume this is always the case for Scorpio.

**Comment by dqwu** [comment]
> Actually, you can simply set `isize` to -1, i.e.
> 
> ```
> ierr = nfmpi_get_var_all(File%fh, varid, ival, -1, {MPITYPE})
> ```
> 
> Using -1 means the number of get elements matches the argument `count`. In this `var` API case, it is the whole variable.
> 
> The requirement is that `MPITYPE` must be the I/O buffer's data type, an MPI predefined datatype, I assume this is always the case for Scorpio.

Thanks for your suggestion. I have tested nfmpi_get_var_all with isize set to -1, and the error still exists (Input/Output data amount mismatch).

**Comment by wkliao** [comment]
I just found out this feature of PnetCDF is only available for C APIs.
So forget about my suggestion.

**Comment by jayeshkrishna** [comment]
This issue is fixed by PR #479
